### ISSUE TITLE
 No value provided for required parameter '_outer_Id'. v2.0.0

### ISSUE THREAD

**Issue by jefridev** [aspnet/EntityFrameworkCore#10122]
Error description:

After upgrading to Entity Framework Core 2.0, It started to throw this  exception about not value provided.

If you are seeing an exception, include the full exceptions details (message and stack trace).

```
Exception message:
Error Message:
 System.InvalidOperationException : No value provided for required parameter '_outer_Idcliente'.
Stack Trace:
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalParameterBase.AddDbParameter(DbCommand command, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.CreateCommand(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, DbCommandMethod executeMethod, IReadOnlyDicti
onary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteReader(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.BufferlessMoveNext(Boolean buffer)
   at Microsoft.EntityFrameworkCore.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded
)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.QueryMethodProvider.<_InjectParameters>d__27`1.MoveNext()
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at System.Collections.Generic.List`1.AddEnumerable(IEnumerable`1 enumerable)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at lambda_method(Closure , QueryContext , ValueBuffer )
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.ProjectionShaper.TypedProjectionShaper`3.Shape(QueryContext queryContext, ValueBuffer va
lueBuffer)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.BufferlessMoveNext(Boolean buffer)
   at Microsoft.EntityFrameworkCore.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded
)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
   at System.Linq.Enumerable.SingleOrDefault[TSource](IEnumerable`1 source)
   at lambda_method(Closure , QueryContext )
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass17_0`1.<CompileQueryCore>b__0(QueryContext qc)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at System.Linq.Queryable.SingleOrDefault[TSource](IQueryable`1 source)
   at QueryProject.Tests.UnitTest.TestQueryWithException() in C:\Users\Jefri Martinez\Desktop\Issues\QueryProject.Tests\UnitTest.cs:line 26
   
```

### Steps to reproduce

I've created this project with a test and scenario for creating the same issue. 
https://github.com/jefridev/IssueEntityFramework20 

Steps to use this project

-  Create database within ScriptsDB Folder.
-  Change connection string in DbContext , QueryProject.Context.SystemContext - OnConfiguring
-  Run the test project, QueryProject.Tests 

[Note]
It was working fine on previous version 1.1.1 but after I upgraded my project to 2.0.0 all queries similar to this one throws the "error no value provided"

### Further technical details
EF Core version: 2.0.0.
Database Provider: (e.g. Microsoft.EntityFrameworkCore.SqlServer)
Operating system:  Windows 7 x64, Windows 10 x64, Mac OS,
IDE: (Visual Studio 2017, VS Code, Rider.NET)

**Comment by ajcvickers** [comment]
@maumar Can you investigate this as a priority as to whether we want to get this into a patch? (Patch dates are being set and we may not have much time.)

**Comment by maumar** [comment]
I was able to reproduce the problem - it happens when the outer parameters is needed during join. We initially create InjectParameter call around the _Join expression, but it gets lost/overwritten during the TryFlattenJoin.

**Comment by ajcvickers** [comment]
@maumar Thanks for investigating. Are there workarounds and if so how easy are they?

**Comment by maumar** [comment]
There is an easy workaround - forcing client evaluation on the join, but it may severely degrade the performance of the query.

Original code (simplified):
cs
```
(from a in ctx.Cliente
select new ClienteViewModel
{
   ListaDireccion =
		(from sd in ctx.ClienteDireccion
		 join s in ctx.Cliente on a.Idcliente equals s.Idcliente // this is the problematic line
		 select new DireccionViewModel
		 {
			 IdDireccion = 42
		 }).ToList()
}).SingleOrDefault()
```

Workaround:

```
public static T ClientMethod<T>(T element) => element;

(...)

(from a in ctx.Cliente
select new ClienteViewModel
{
   ListaDireccion =
		(from sd in ctx.ClienteDireccion
		 join s in ctx.Cliente on ClientMethod(a.Idcliente) equals s.Idcliente
		 select new DireccionViewModel
		 {
			 IdDireccion = 42
		 }).ToList()
}).SingleOrDefault()
```

**Comment by maumar** [comment]
fixed in 12e5c1b58211c18bbbf722dc1a4cf67d8aa81806
